### PR TITLE
TimeSeriesPanel: Null points no longer get tooltips

### PR DESCRIPTION
**Pull request by kaydelaney**
**What this PR does / why we need it**:
There was a small issue where hovering near a point on a time series panel with multiple series could result in null values receiving tooltips, depending on the specific alignment of points in the series being displayed.

https://user-images.githubusercontent.com/45561153/143590064-09579626-4991-429a-8b6e-fff92156a51c.mp4

After using `dataIdx` instead of `idx`:

https://user-images.githubusercontent.com/45561153/143590079-cb796877-d433-475e-adc1-c3c9fae925d4.mp4

### CODE REVIEW

File: packages/grafana-ui/src/components/uPlot/plugins/TooltipPlugin.tsx [hunk 1]
```diff
@@ -174,19 +172,28 @@ export const TooltipPlugin: React.FC<TooltipPluginProps> = ({
...
     if (mode === TooltipDisplayMode.Single && focusedSeriesIdx !== null) {
       const field = otherProps.data.fields[focusedSeriesIdx];
+      const xValueAtCursor = plotInstance.current?.posToVal(plotInstance.current?.cursor.left!, 'x');
+      const dataIdx =
+        plotInstance.current?.cursor.dataIdx?.(
+          plotInstance.current,
+          focusedSeriesIdx,
+          focusedPointIdx,
+          xValueAtCursor ?? 0
+        ) ?? focusedPointIdx;
```
Comment (leeoniya) [inline packages/grafana-ui/src/components/uPlot/plugins/TooltipPlugin.tsx:188]: i _think_ this can be simplified by accessing `plotInstance.current?.legend.idxs`. which is a series-matched array of current `dataIdxs`.

can you add a panel with data that repos this to http://localhost:3000/d/8mmCAF1Mz/panel-tests-graph-ng-gaps-and-connected?orgId=1. i'd like to try it out.

Comment (leeoniya) [reply packages/grafana-ui/src/components/uPlot/plugins/TooltipPlugin.tsx:188]: also, i still need to port over the fixes to the hover null-scanning code from https://github.com/leeoniya/uPlot/commit/53f2b179ed5647218e507b7e73c183ec4157b207 :see_no_evil:

Comment (kaydelaney) [reply packages/grafana-ui/src/components/uPlot/plugins/TooltipPlugin.tsx:188]: The following Panel JSON repros the issue:
```
{
  "id": 4,
  "gridPos": {
    "h": 8,
    "w": 12,
    "x": 0,
    "y": 0
  },
  "type": "timeseries",
  "title": "Panel Title",
  "fieldConfig": {
    "defaults": {
      "custom": {
        "drawStyle": "points",
        "lineInterpolation": "linear",
        "barAlignment": 0,
        "lineWidth": 1,
        "fillOpacity": 0,
        "gradientMode": "none",
        "spanNulls": false,
        "showPoints": "auto",
        "pointSize": 30,
        "stacking": {
          "mode": "none",
          "group": "A"
        },
        "axisPlacement": "auto",
        "axisLabel": "",
        "scaleDistribution": {
          "type": "linear"
        },
        "hideFrom": {
          "tooltip": false,
          "viz": false,
          "legend": false
        },
        "thresholdsStyle": {
          "mode": "off"
        }
      },
      "color": {
        "mode": "palette-classic"
      },
      "thresholds": {
        "mode": "absolute",
        "steps": [
          {
            "color": "green",
            "value": null
          },
          {
            "color": "red",
            "value": 80
          }
        ]
      },
      "mappings": []
    },
    "overrides": []
  },
  "options": {
    "tooltip": {
      "mode": "single"
    },
    "legend": {
      "displayMode": "list",
      "placement": "bottom",
      "calcs": []
    }
  },
  "targets": [
    {
      "datasource": {
        "type": "testdata",
        "uid": "PD8C576611E62080A"
      },
      "refId": "A",
      "scenarioId": "csv_metric_values",
      "stringInput": "1,20,90,30,5,0"
    },
    {
      "refId": "B",
      "scenarioId": "csv_metric_values",
      "stringInput": "1,1,1,1,1,1,1,1"
    }
  ],
  "datasource": null
}
```

If you hover your cursor towards the bottom right of the top-most point, you'll see the tooltip pop up with no data like in the video.

Comment (leeoniya) [reply packages/grafana-ui/src/components/uPlot/plugins/TooltipPlugin.tsx]: thanks!

my test dashboard for this is:

```json
{
  "annotations": {
    "list": [
      {
        "builtIn": 1,
        "datasource": "-- Grafana --",
        "enable": true,
        "hide": true,
        "iconColor": "rgba(0, 211, 255, 1)",
        "name": "Annotations & Alerts",
        "target": {
          "limit": 100,
          "matchAny": false,
          "tags": [],
          "type": "dashboard"
        },
        "type": "dashboard"
      }
    ]
  },
  "editable": true,
  "fiscalYearStartMonth": 0,
  "graphTooltip": 0,
  "id": 224,
  "links": [],
  "liveNow": false,
  "panels": [
    {
      "fieldConfig": {
        "defaults": {
          "color": {
            "mode": "palette-classic"
          },
          "custom": {
            "axisLabel": "",
            "axisPlacement": "auto",
            "barAlignment": 0,
            "drawStyle": "points",
            "fillOpacity": 0,
            "gradientMode": "none",
            "hideFrom": {
              "legend": false,
              "tooltip": false,
              "viz": false
            },
            "lineInterpolation": "linear",
            "lineWidth": 1,
            "pointSize": 5,
            "scaleDistribution": {
              "type": "linear"
            },
            "showPoints": "auto",
            "spanNulls": false,
            "stacking": {
              "group": "A",
              "mode": "none"
            },
            "thresholdsStyle": {
              "mode": "off"
            }
          },
          "mappings": [],
          "thresholds": {
            "mode": "absolute",
            "steps": [
              {
                "color": "green",
                "value": null
              },
              {
                "color": "red",
                "value": 80
              }
            ]
          }
        },
        "overrides": []
      },
      "gridPos": {
        "h": 8,
        "w": 8,
        "x": 0,
        "y": 0
      },
      "id": 6,
      "options": {
        "legend": {
          "calcs": [],
          "displayMode": "list",
          "placement": "bottom"
        },
        "tooltip": {
          "mode": "single"
        }
      },
      "targets": [
        {
          "refId": "A",
          "scenarioId": "csv_metric_values",
          "stringInput": "1,20,90,30,5,0"
        },
        {
          "refId": "B",
          "scenarioId": "csv_metric_values",
          "stringInput": "1,1,1,1,1,1,1,1"
        }
      ],
      "title": "Panel Title",
      "type": "timeseries"
    },
    {
      "fieldConfig": {
        "defaults": {
          "color": {
            "mode": "palette-classic"
          },
          "custom": {
            "axisLabel": "",
            "axisPlacement": "auto",
            "barAlignment": 0,
            "drawStyle": "line",
            "fillOpacity": 0,
            "gradientMode": "none",
            "hideFrom": {
              "legend": false,
              "tooltip": false,
              "viz": false
            },
            "lineInterpolation": "linear",
            "lineWidth": 1,
            "pointSize": 5,
            "scaleDistribution": {
              "type": "linear"
            },
            "showPoints": "auto",
            "spanNulls": false,
            "stacking": {
              "group": "A",
              "mode": "none"
            },
            "thresholdsStyle": {
              "mode": "off"
            }
          },
          "mappings": [],
          "thresholds": {
            "mode": "absolute",
            "steps": [
              {
                "color": "green",
                "value": null
              },
              {
                "color": "red",
                "value": 80
              }
            ]
          }
        },
        "overrides": []
      },
      "gridPos": {
        "h": 8,
        "w": 12,
        "x": 0,
        "y": 8
      },
      "id": 4,
      "options": {
        "legend": {
          "calcs": [],
          "displayMode": "list",
          "placement": "bottom"
        },
        "tooltip": {
          "mode": "single"
        }
      },
      "targets": [
        {
          "refId": "A",
          "scenarioId": "csv_metric_values",
          "stringInput": "2,3,4,null"
        },
        {
          "refId": "B",
          "scenarioId": "csv_metric_values",
          "stringInput": "1,20,90,30,5,0"
        }
      ],
      "title": "Panel Title",
      "transformations": [
        {
          "id": "calculateField",
          "options": {
            "binary": {
              "left": "A-series",
              "operator": "-",
              "reducer": "sum",
              "right": "5"
            },
            "mode": "binary",
            "reduce": {
              "reducer": "sum"
            }
          }
        }
      ],
      "type": "timeseries"
    },
    {
      "fieldConfig": {
        "defaults": {
          "color": {
            "mode": "palette-classic"
          },
          "custom": {
            "axisLabel": "",
            "axisPlacement": "auto",
            "barAlignment": 0,
            "drawStyle": "line",
            "fillOpacity": 0,
            "gradientMode": "none",
            "hideFrom": {
              "graph": false,
              "legend": false,
              "tooltip": false,
              "viz": false
            },
            "lineInterpolation": "linear",
            "lineWidth": 1,
            "pointSize": 10,
            "scaleDistribution": {
              "type": "linear"
            },
            "showPoints": "auto",
            "spanNulls": true,
            "stacking": {
              "group": "A",
              "mode": "none"
            },
            "thresholdsStyle": {
              "mode": "off"
            }
          },
          "mappings": [],
          "thresholds": {
            "mode": "absolute",
            "steps": [
              {
                "color": "green",
                "value": null
              },
              {
                "color": "red",
                "value": 80
              }
            ]
          }
        },
        "overrides": []
      },
      "gridPos": {
        "h": 9,
        "w": 12,
        "x": 0,
        "y": 16
      },
      "id": 2,
      "options": {
        "legend": {
          "calcs": [],
          "displayMode": "list",
          "placement": "bottom"
        },
        "tooltip": {
          "mode": "multi"
        }
      },
      "targets": [
        {
          "refId": "A",
          "scenarioId": "csv_metric_values",
          "stringInput": "30,null,null,null,null,null,null,30,null,null,null,null,null,30"
        },
        {
          "refId": "B",
          "scenarioId": "csv_metric_values",
          "stringInput": "5,5,5,5,5,5,5,5,5,5,5,5,5,5,5,5,5,5,5,5,5,5,5,5,5,5,5,5,5,5,5,5,5"
        },
        {
          "refId": "C",
          "scenarioId": "csv_metric_values",
          "stringInput": "18,18,18,18,18,18,18,18,18,18,18,18,18,18,18,18,18,18,18,18,18,18,18,18,18,18,18,18,18,18,18,18"
        },
        {
          "refId": "D",
          "scenarioId": "csv_metric_values",
          "stringInput": "25,25,25,25,25,25,25,25,25,25,25,25,25,25,25,25,25,25,25,25,25,25,25,25,25,25,25,25,25,25,25"
        }
      ],
      "title": "Panel Title",
      "type": "timeseries"
    }
  ],
  "refresh": "",
  "schemaVersion": 33,
  "style": "dark",
  "tags": [],
  "templating": {
    "list": []
  },
  "time": {
    "from": "now-6h",
    "to": "now"
  },
  "timepicker": {},
  "timezone": "",
  "title": "null-nearest",
  "uid": "iaqi4reGz",
  "version": 16,
  "weekStart": ""
}
```

File: packages/grafana-ui/src/components/uPlot/plugins/TooltipPlugin.tsx [hunk 2]
```diff
@@ -233,7 +233,7 @@ export const TooltipPlugin: React.FC<TooltipPluginProps> = ({
       tooltip = <SeriesTable series={series} timestamp={xVal} />;
     }
   } else {
-    tooltip = renderTooltip(otherProps.data, focusedSeriesIdx, focusedPointIdx);
+    tooltip = renderTooltip(otherProps.data, focusedSeriesIdx, plotInstance.current?.cursor.idx ?? focusedPointIdx);
```
Comment (leeoniya) [inline packages/grafana-ui/src/components/uPlot/plugins/TooltipPlugin.tsx:236]: this looks unrelated? (and breaks State timeline tooltip)

http://localhost:3000/d/mIJjFy8Gz/timeline-modes?orgId=1

**Comment by leeoniya** [comment]
i've made a couple followup commits here: https://github.com/grafana/grafana/pull/42397

feel free to cherrypick and close that PR out.

**Comment by kaydelaney** [comment]
>  Should this PR expand to include the multi-series tooltip mode and filter out null values?

Honestly not sure - the UX might be a bit annoying if we filter out null values, with the displayed series/values constantly changing as the mouse is moved around.

https://user-images.githubusercontent.com/45561153/143906367-83cb910e-da8c-438d-9e67-77018fdd06e3.mp4

Perhaps best to defer to the BI squad's judgement here. Thoughts @dprokop @jackw ?

**Review by zoltanbedi — APPROVED** [review]
I've tried this out and works as expected. My personal opinion is that we shouldn't filter the null values from the tooltip. Maybe we should bring this up in the ux feedback session?

**Comment by dprokop** [comment]
> My personal opinion is that we shouldn't filter the null values from the tooltip.

I agree with @zoltanbedi. If those were to be hidden I would make in an opt-in feature rather than a default behavior.
